File: modules/vision-camera-frame-diff/ios/FrameDiffPlugin.swift
import VisionCamera

/**
 * VisionCamera frame processor plugin that computes luminance-based
 * frame differencing on each camera frame.
 *
 * Registered as "frameDiff" — called from JS via:
 *   `VisionCameraProxy.initFrameProcessorPlugin('frameDiff')`
 *
 * Returns a Double (0-1 motion magnitude), or nil on first frame.
 *
 * Also stores the latest result in a thread-safe static property so the
 * Expo module can expose it to JS via synchronous polling.
 */
@objc(FrameDiffPlugin)
class FrameDiffPlugin: FrameProcessorPlugin {

  private let computer = FrameDiffComputer()

  // Thread-safe storage for latest motion magnitude.
  // Written from the frame processor thread, read from JS thread via Expo module.
  private static let lock = NSLock()
  private static var _latestMotion: Double?

  static var latestMotion: Double? {
    get {
      lock.lock()
      defer { lock.unlock() }
      return _latestMotion
    }
    set {
      lock.lock()
      defer { lock.unlock() }
      _latestMotion = newValue
    }
  }

  override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]? = nil) {
    super.init(proxy: proxy, options: options)
  }

  override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any? {
    let buffer = frame.buffer
    let result = computer.computeDiff(sampleBuffer: buffer)

    FrameDiffPlugin.latestMotion = result
    return result
  }
}

File: modules/vision-camera-frame-diff/android/src/main/java/com/divotgolf/framediff/FrameDiffPlugin.kt
package com.divotgolf.framediff

import android.util.Log
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin

/**
 * VisionCamera frame processor plugin that computes luminance-based
 * frame differencing on each camera frame.
 *
 * Registered as "frameDiff" from [VisionCameraFrameDiffModule].
 * Called from JS via: `VisionCameraProxy.initFrameProcessorPlugin('frameDiff')`
 *
 * Returns a Double (0-1 motion magnitude), or null on first frame.
 */
class FrameDiffPlugin : FrameProcessorPlugin() {

  private val computer = FrameDiffComputer()

  init {
    Log.d(TAG, "FrameDiffPlugin instance created")
  }

  override fun callback(frame: Frame, arguments: Map<String, Any>?): Any? {
    val image = frame.image
    val result = computer.computeDiff(image)

    latestMotion = result

    if (frameCount % LOG_INTERVAL == 0L && result != null) {
      Log.d(TAG, "Frame #$frameCount: motion=%.4f".format(result))
    }
    frameCount++

    return result
  }

  companion object {
    private const val TAG = "FrameDiff"
    private const val LOG_INTERVAL = 60L
    private var frameCount = 0L

    /** Latest motion magnitude. Written from frame processor thread, read from JS thread. */
    @Volatile
    var latestMotion: Double? = null
  }
}
